src/boot/efi/initrd.c

```diff
@@ -111,11 +127,18 @@ EFI_STATUS initrd_unregister(InitrdLoader *loader) {
         if (!loader)
                 return EFI_SUCCESS;
 
-        err = BS->UninstallMultipleProtocolInterfaces(
-                        loader->dp_handle,
-                        MAKE_GUID_PTR(EFI_DEVICE_PATH_PROTOCOL), &efi_initrd_device_path,
-                        MAKE_GUID_PTR(EFI_LOAD_FILE2_PROTOCOL), loader,
-                        NULL);
+        if (loader->image_handle)
+                err = BS->UninstallMultipleProtocolInterfaces(
+                                loader->image_handle,
+                                MAKE_GUID_PTR(LINUX_EFI_INITRD_LF2_PROTOCOL), loader,
+                                NULL);
+        else
+                err = BS->UninstallMultipleProtocolInterfaces(
+                                loader->dp_handle,
+                                MAKE_GUID_PTR(EFI_DEVICE_PATH_PROTOCOL), &efi_initrd_device_path,
+                                MAKE_GUID_PTR(EFI_LOAD_FILE2_PROTOCOL), loader,
+                                NULL);
+
```

Comment:
nitpick: we try to keep funciton calls and there error handling together, hence maybe drop this empty line